contrib/fuzz/container_fuzzer.go

```diff
@@ -0,0 +1,320 @@
+// +build gofuzz
+
+/*
+   Copyright The containerd Authors.
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+       http://www.apache.org/licenses/LICENSE-2.0
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+/*
+	To run this fuzzer, it must first be moved to
+	integration/client. OSS-fuzz does this automatically
+	everytime it builds the fuzzers.
+*/
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	fuzz "github.com/AdaLogics/go-fuzz-headers"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/oci"
+	"github.com/containerd/containerd/sys"
+)
+
+func init() {
+	err := updatePathEnv()
+	if err != nil {
+		panic(err)
+	}
+}
+
+func tearDown() error {
+	if err := ctrd.Stop(); err != nil {
+		if err := ctrd.Kill(); err != nil {
+			return err
+		}
+	}
+	if err := ctrd.Wait(); err != nil {
+		if _, ok := err.(*exec.ExitError); !ok {
+			return err
+		}
+	}
+	if err := sys.ForceRemoveAll(defaultRoot); err != nil {
+		return err
+	}
+	return nil
+}
+
+// startDaemon() starts the daemon.
+func startDaemon(ctx context.Context) {
+	buf := bytes.NewBuffer(nil)
+	stdioFile, err := ioutil.TempFile("", "")
+	if err != nil {
+		// We panic here as it is a fuzz-blocker that
+		// may need fixing
+		panic(err)
+	}
+	defer func() {
+		stdioFile.Close()
+		os.Remove(stdioFile.Name())
+	}()
+	ctrdStdioFilePath = stdioFile.Name()
+	stdioWriter := io.MultiWriter(stdioFile, buf)
+	err = ctrd.start("containerd", address, []string{
+		"--root", defaultRoot,
+		"--state", defaultState,
+		"--log-level", "debug",
+		"--config", createShimDebugConfig(),
+	}, stdioWriter, stdioWriter)
+	if err != nil {
+		// We are fine if the error is that the daemon is already running,
+		// but if the error is another, then it will be a fuzz blocker,
+		// so we panic
+		if !strings.Contains(err.Error(), "daemon is already running") {
+			fmt.Fprintf(os.Stderr, "%s: %s\n", err, buf.String())
+			panic(err)
+		}
+	}
+	// Whether or not we should tear down after each fuzz iteration
+	// requires further investigation. The fuzzer runs fine without
+	// but if there is no performance-penalty, then we should tear down.
+	// To-do @AdamKorcz
+	/*defer func() {
+		err = tearDown()
+		if err != nil {
+			panic(err)
+		}
+	}()*/
```

Comment:
I think it is better to start from slow-but-better method. In this case, it would be better to have this tearDown method first.

Reply:
I suggest we then have two versions of this fuzzer to measure each fuzzers performance: One that tears down and one that doesn't.
I can add the second version and comments that explain it.

Reply:
I have added a version of the fuzzer that tears down. I suggest we run both for now.